File: android/app/src/main/java/com/coinglass/intel/domain/Symbols.kt
package com.coinglass.intel.domain

data class SymbolInfo(
    val symbol: String,
    val base: String,
    val quote: String = "USDT",
    val binance: String,
    val bybit: String,
    val okx: String,
)

object Symbols {
    val registry: Map<String, SymbolInfo> = mapOf(
        "BTCUSDT" to info("BTC"),
        "ETHUSDT" to info("ETH"),
        "SOLUSDT" to info("SOL"),
        "EDGEUSDT" to info("EDGE"),
        "ALLOUSDT" to info("ALLO"),
        "SPELLUSDT" to info("SPELL"),
        "XAUTUSDT" to info("XAUT"),
        "BLUAIUSDT" to info("BLUAI"),
    )

    val chips: List<String> = listOf(
        "BTCUSDT", "ETHUSDT", "SOLUSDT", "ALLOUSDT", "EDGEUSDT", "SPELLUSDT",
    )

    fun resolve(raw: String): SymbolInfo {
        val s = normalize(raw)
        return registry[s] ?: info(s.removeSuffix("USDT")).copy(symbol = s, binance = s, bybit = s)
    }

    fun normalize(raw: String): String {
        var s = raw.uppercase()
            .replace("-PERP", "")
            .replace("_PERP", "")
            .replace("-", "")
            .replace("/", "")
            .trim()
        if (s.isEmpty()) return "BTCUSDT"
        if (s.endsWith("USDT") || s.endsWith("USDC")) return s
        if (s.endsWith("USD")) return s
        return s + "USDT"
    }

    fun base(raw: String): String {
        val s = normalize(raw)
        return when {
            s.endsWith("USDT") -> s.dropLast(4)
            s.endsWith("USDC") -> s.dropLast(4)
            s.endsWith("USD") -> s.dropLast(3)
            else -> s
        }
    }

    private fun info(base: String): SymbolInfo = SymbolInfo(
        symbol = base + "USDT",
        base = base,
        binance = base + "USDT",
        bybit = base + "USDT",
        okx = "$base-USDT-SWAP",
    )
}

File: engine/symbols.py
"""Known futures symbol map (scalper registry). Unknown symbols still work via convention."""
from __future__ import annotations

from typing import Dict

SYMBOLS_REGISTRY: Dict[str, Dict[str, str]] = {
    "EDGEUSDT": {
        "base": "EDGE", "quote": "USDT",
        "coingecko_id": "edgex", "coinpaprika_id": "edge-edgex",
        "binance": "EDGEUSDT", "bybit": "EDGEUSDT", "okx": "EDGE-USDT-SWAP",
    },
    "ALLOUSDT": {
        "base": "ALLO", "quote": "USDT",
        "coingecko_id": "allora", "coinpaprika_id": "allora",
        "binance": "ALLOUSDT", "bybit": "ALLOUSDT", "okx": "ALLO-USDT-SWAP",
    },
    "SPELLUSDT": {
        "base": "SPELL", "quote": "USDT",
        "coingecko_id": "spell-token", "coinpaprika_id": "spell-token",
        "binance": "SPELLUSDT", "bybit": "SPELLUSDT", "okx": "SPELL-USDT-SWAP",
    },
    "XAUTUSDT": {
        "base": "XAUT", "quote": "USDT",
        "coingecko_id": "tether-gold", "coinpaprika_id": "xaut-tether-gold",
        "binance": "XAUTUSDT", "bybit": "XAUTUSDT", "okx": "XAUT-USDT-SWAP",
    },
    "BLUAIUSDT": {
        "base": "BLUAI", "quote": "USDT",
        "coingecko_id": "blueai", "coinpaprika_id": "blueai",
        "binance": "BLUAIUSDT", "bybit": "BLUAIUSDT", "okx": "BLUAI-USDT-SWAP",
    },
}


def resolve_symbol(raw: str) -> Dict[str, str]:
    s = raw.upper().replace("-PERP", "").replace("_PERP", "").replace("-", "")
    if not s.endswith("USDT"):
        s = f"{s}USDT"
    if s in SYMBOLS_REGISTRY:
        return {"symbol": s, **SYMBOLS_REGISTRY[s]}
    base = s.replace("USDT", "")
    return {
        "symbol": s, "base": base, "quote": "USDT",
        "coingecko_id": base.lower(), "coinpaprika_id": base.lower(),
        "binance": s, "bybit": s, "okx": f"{base}-USDT-SWAP",
    }
